### PR TITLE
create multi queue veth for xdp

### PR DESCRIPTION
**Pull request by arikachen**
Signed-off-by: chenqijun <chenqijun@corp.netease.com>

## Description

Now we use vpp + vcl to accelerate apps in containers, using xdp/afxdp to redirect packet from veth to vpp. 
vpp has more than one workers to process packet in our situation,  
only one vpp worker works when veth has one queue, it affect some performance.

It depend more recent kernel support.

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

**Comment by caseydavenport** [comment]
@arikachen thanks for the PR - could you add some context to the description explaining your use-case?

/sem-approve

File: pkg/dataplane/linux/dataplane_linux.go [hunk 1]
```diff
@@ -76,8 +78,10 @@ func (d *linuxDataplane) DoNetworking(
 	err = ns.WithNetNSPath(args.Netns, func(hostNS ns.NetNS) error {
 		veth := &netlink.Veth{
 			LinkAttrs: netlink.LinkAttrs{
-				Name: contVethName,
-				MTU:  d.mtu,
+				Name:        contVethName,
+				MTU:         d.mtu,
+				NumTxQueues: d.queues,
+				NumRxQueues: d.queues,
```
Comment (neiljerram) [inline pkg/dataplane/linux/dataplane_linux.go:84]: What was the previous default for these fields, when we weren't setting them explicitly?

I guess it was 1, and so by default this PR doesn't change that, but adds the capability to do that by configuring `n_queues`.  Is that right?

**Review by neiljerram — COMMENTED** [review]
Code looks good, but like @caseydavenport I'd also like more description of the motivation for this.  I also dropped one specific comment.

**Comment by arikachen** [comment]
@caseydavenport @neiljerram 
thanks for you reply.  I update description in pr.

File: pkg/dataplane/linux/dataplane_linux.go (hunk 1 above)
Comment (arikachen) [reply pkg/dataplane/linux/dataplane_linux.go:84]: yes

**Comment by neiljerram** [comment]
Thanks @arikachen .  I guess the XDP -> VPP path only works for the ingress direction.  Does that mean we should only change NumRxQueues?

**Comment by arikachen** [comment]
> Thanks @arikachen . I guess the XDP -> VPP path only works for the ingress direction. Does that mean we should only change NumRxQueues?

Yes, it is. but kernel limit the queue num with peer tx queue num.
```
if (dev->real_num_rx_queues < peer->real_num_tx_queues) {
            NL_SET_ERR_MSG_MOD(extack, "XDP expects number of rx queues not less than peer tx queues");
            err = -ENOSPC;
            goto err;
        }
```

**Comment by arikachen** [comment]
@neiljerram Thanks.  I have test with increase rx queue,  only one rx queue works.
 ```
# ethtool -S eth0
NIC statistics:
     peer_ifindex: 351
     rx_queue_0_xdp_packets: 10570419
     rx_queue_0_xdp_bytes: 1526253129
     rx_queue_0_xdp_drops: 0
     rx_queue_1_xdp_packets: 0
     rx_queue_1_xdp_bytes: 0
     rx_queue_1_xdp_drops: 0
# ethtool -l eth0
Channel parameters for eth0:
Pre-set maximums:
RX:             2
TX:             1
Other:          0
Combined:       1
Current hardware settings:
RX:             2
TX:             1
Other:          0
Combined:       1
```

**Comment by neiljerram** [comment]
@arikachen Thanks for your last explanatory comment.  Let's see if we can get the CI to pass now...

**Review by neiljerram — APPROVED** [review]
LGTM - @caseydavenport do you have any concerns?

File: pkg/types/types.go [hunk 2]
```diff
@@ -80,6 +80,7 @@ type NetConf struct {
 	} `json:"ipam,omitempty"`
 	Args                 Args                   `json:"args"`
 	MTU                  int                    `json:"mtu"`
+	NumQueues            int                    `json:"n_queues"`
```
Comment (neiljerram) [inline pkg/types/types.go:83]: One of my colleagues has asked if we can change this JSON name to "num_queues", for a little more immediate clarity.

**Review by neiljerram — CHANGES_REQUESTED** [review]
@arikachen Would you mind making one small naming change?  Then I think we can merge this.

**Comment by arikachen** [comment]
> @arikachen Would you mind making one small naming change? Then I think we can merge this.

@neiljerram thanks.

**Review by neiljerram — APPROVED** [review]

**Pull request merged**